stl/inc/xutility

```diff
@@ -4180,7 +4180,316 @@ _NODISCARD _CONSTEXPR17 move_iterator<_Iter> make_move_iterator(_Iter _It) { //
+        const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Left._Mystate != _State::_Nonthing && _Right._Mystate != _State::_Nonthing,
+            "Cannot take difference of valueless common_iterators");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter) {
+            return _Left._Myiter - _Right._Myiter;
+        } else if (_Left._Mystate == _State::_Holds_sentinel && _Right._Mystate == _State::_Holds_sentinel) {
+            return 0;
+        } else if (_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_sentinel) {
+            return _Left._Myiter - _Right._Mysentinel;
+        } else {
+            return _Left._Mysentinel - _Right._Myiter;
+        }
+    }
+
+    _NODISCARD friend iter_rvalue_reference_t<_Iter> iter_move(const common_iterator& _Right) noexcept(
+        noexcept(_RANGES iter_move(_STD declval<const _Iter&>())))
+        requires input_iterator<_Iter>
+    {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(
+            _Right._Mystate == _State::_Holds_iter, "Can only iter_move from common_iterator if it holds an interator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _RANGES iter_move(_Right._Myiter);
+    }
+    // clang-format on
+
+    template <indirectly_swappable<_Iter> _Iter2, class _Se2>
+    friend void iter_swap(const common_iterator& _Left, const common_iterator<_Iter2, _Se2>& _Right) noexcept(
+        noexcept(_RANGES iter_swap(_STD declval<const _Iter&>(), _STD declval<const _Iter2&>()))) {
+#if _ITERATOR_DEBUG_LEVEL != 0
+        _STL_VERIFY(_Left._Mystate == _State::_Holds_iter && _Right._Mystate == _State::_Holds_iter,
+            "Can only iter_swap from common_iterators if both hold an interator");
+#endif // _ITERATOR_DEBUG_LEVEL != 0
+        return _RANGES iter_swap(_Left._Myiter, _Right._Myiter);
+    }
+
+private:
+    enum class _State : unsigned char { _Nonthing, _Holds_iter, _Holds_sentinel };
+
+    union {
+        _Iter _Myiter;
+        _Se _Mysentinel;
+        _Nontrivial_dummy_type _Mydummy;
+    };
+    _State _Mystate;
+};
+
+template <class _Iter, class _Se>
+struct incrementable_traits<common_iterator<_Iter, _Se>> {
+    using difference_type = iter_difference_t<_Iter>;
+};
+
+template <input_iterator _Iter, class _Se>
+struct iterator_traits<common_iterator<_Iter, _Se>> {
+    using iterator_concept = conditional_t<forward_iterator<_Iter>, forward_iterator_tag, input_iterator_tag>;
+    using iterator_category =
+        conditional_t<derived_from<_Iter_cat_t<_Iter>, forward_iterator_tag>, forward_iterator_tag, input_iterator_tag>;
+    using value_type      = iter_value_t<_Iter>;
+    using difference_type = iter_difference_t<_Iter>;
+    using pointer         = conditional_t<_Has_op_arrow<_Iter>, decltype(_STD declval<_Iter&>().operator->()), void>;

```

Comment:
Let me guess, for that to be well formed it needs operator->() to actually exist and I need an indirection again?